pkg/storage/segments/store.go

```diff
@@ -76,30 +77,56 @@ func (s *segmentStore) Put(ctx context.Context, path paths.Path, data io.Reader,
 	metadata []byte, expiration time.Time) (meta Meta, err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	// uses overlay client to request a list of nodes
-	nodes, err := s.oc.Choose(ctx, s.rs.TotalCount(), 0)
+	var p *ppb.Pointer
+
+	// checks if pointer should be inline or remote
+	// TODO(nat): check that size is less than s.thresholdSize
+	if data == nil {
```

Comment:
Is this condition `data == nil` just a placeholder until you implement the logic to check that size is less than `s.thresholdSize`?

Reply:
So here where the major challenge is. The segment store needs to have some sense of how much data (more or less than the `thresholdSize`) is available in the given `data` reader.  But this check should be executed without taking the data out of the reader. Because this same `data` reader will be potentially given to the ECClient, which needs to read it from the very beginning, i.e. without losing these first bytes.

One possible solution is to implement a wrapper PeekThresholdReader type, which will:
  - Take an `io.Reader` as parameter in the constructor
  - Implement the `io.Reader` interface itself
  - Provide an additional method `checkThreshold` which will try to read `thresholdSize` bytes from the wrapped reader into an internal buffer.

Then you can use it like this:
```
peekReader := PeekThresholdReader(data)
bigEnough := peekReader.checkThreshold(thresholdSize)
if (bigEnough) {
  // remote segment
} else {
  // internal segment
}
```
Then any data for the internal or remote segment should be read from the PeekThresholdReader instead from the `data` reader. The Read method in the PeekThresholdReader will be implemented in a way that it first reads the bytes from that internal buffer and then continues with reading from the wrapped `data` reader. No more buffering from this moment on.